[2.11.x] DDF-3604 Fixed OpenSearch bounding box

emmberk opened the pull request:
#### What does this PR do?
This PR fixes the OpenSearch `Convert to BBox` option.
#### Who is reviewing it? 
@AzGoalie @peterhuffer 
#### Select relevant component teams: 
@codice/io 
#### Choose 2 committers to review/merge the PR. 
@bdeining
@clockard
#### How should this be tested?
Execute point radius searches to an OpenSearch source in Intrigue, and confirm that the expected results are returned.
#### Any background context you want to provide?
Specifically, the issue was observed with a record with `location`=`POLYGON ((-122 37.33, -122.05 37.33, -122.05 37.22, -122 37.22, -122 37.33))` and the search "`Latitude`=`33`, `Longitude`=`-122`, `Radius`=`500 miles`" from Intrigue. The `OpenSearchSource`, with `Convert to BBox`=`true`, was incorrectly transforming the spatial search to `bbox=422.42053621382615,25.77186615764653,-666.4205362138262,40.22813384235347`.
#### What are the relevant tickets?
[DDF-3604](https://codice.atlassian.net/browse/DDF-3604)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

bdeining commented on catalog/opensearch/catalog-opensearch-source/src/test/java/ddf/catalog/source/opensearch/impl/TestOpenSearchParserImpl.java:
```diff
@@ -69,16 +66,16 @@
   @Before
   public void setUp() {
     openSearchParserImpl = new OpenSearchParserImpl();
+    webClient = WebClient.create("http://localhost:8080/services/catalog/query");
```
✏️ i wonder if we should just use `http://www.example.com`

clockard commented on catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchParserImpl.java:
```diff
@@ -341,34 +341,42 @@ private boolean hasParameter(String parameter, List<String> parameters) {
-    if (maxX > MAX_LON) {
-      maxX -= MAX_ROTATION;
-    }
+    final double latDifference = Math.toDegrees(radius / RADIUS_OF_THE_EARTH_IN_METERS);
```
This method takes in WGS-84 coordinates but then makes the calculation treating the earth as a sphere. How much to we care about our precision here?

emmberk replied:
I haven't thought about this. I'll do some more research and test with both the 2D and 3D maps.

emmberk replied:
There is a disclaimer in the documentation:
https://github.com/codice/ddf/blob/2c280fca47752d4f0fe573827e815def461fd02d/distribution/docs/src/main/resources/content/_tables/OpenSearchSource.adoc#L61-L64

clockard replied:
Good enough for me.

peterhuffer commented on catalog/opensearch/catalog-opensearch-source/src/test/groovy/ddf/catalog/source/opensearch/impl/TestOpenSearchParserImplCreateBBox.groovy:
```diff
@@ -0,0 +1,53 @@
+        211  | 211  | 1      || -180                | 210.99999100678394 | 180                 | 90
+    }
+
+    def 'create BBox from polygon'(String[] polyAry, expectedWest, expectedSouth, expectedEast, expectedNorth) {
```
❓ Is this test cases parameters required?

emmberk replied:
No they're not. I like it this way, though, because it's easy to add more cases in the future and it matched the first method.

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/test/java/ddf/catalog/source/opensearch/impl/TestOpenSearchParserImpl.java:
```diff
@@ -94,9 +91,6 @@ public void populateSpatialDistanceFilterBbox() throws Exception {
-    String urlString = URL.toString();
-    assertThat(urlString, containsString(OpenSearchParserImpl.SEARCH_TERMS));
-    WebClient webClient = WebClient.create(URL.toString());
     SpatialDistanceFilter spatialFilter = new SpatialDistanceFilter("POINT (-211 -211)", 1.0);
```
Aren't these lon and lat values used in a few of these unit tests invalid?

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/test/java/ddf/catalog/source/opensearch/impl/OpenSearchParserImplTest.java:
```diff
@@ -294,7 +151,6 @@ public void populateContextualEmptyMap() {
                 .split(",")));
     String urlStr = webClient.getCurrentURI().toString();
     assertThat(urlStr, not(containsString("q=")));
-    assertThat(urlStr, containsString(OpenSearchParserImpl.SEARCH_TERMS));
```
These missing search terms will be fixed in the next release. I'll add the ticket number here when I find it.

bdeining replied:
@emmberk did you find that ticket number

emmberk replied:
Not yet. I'm working on it now.

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/test/java/ddf/catalog/source/opensearch/impl/OpenSearchParserImplTest.java:
```diff
@@ -69,141 +66,16 @@
-  }
-
-  @Test
-  public void populateSpatialDistanceFilterBbox() throws Exception {
```
Moved these tests to the new groovy test class

emmberk commented:
I made some updates to the unit tests. This PR should be reviewed again.

rzwiefel commented on catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchParserImpl.java:
```diff
@@ -341,34 +341,42 @@ private boolean hasParameter(String parameter, List<String> parameters) {
+    final double latDifference = Math.toDegrees(radius / RADIUS_OF_THE_EARTH_IN_METERS);
+    final double lonDifference =
+        Math.toDegrees(
+            Math.asin(Math.sin(Math.toRadians(latDifference)) / Math.cos(Math.toRadians(lat))));
```
Here is outline of solving it out
![img_20180208_095428](https://user-images.githubusercontent.com/9013416/35987135-2ade0e26-0cb8-11e8-86a0-76c36a085e14.jpg)
![img_20180208_095438](https://user-images.githubusercontent.com/9013416/35987143-2ec74db8-0cb8-11e8-8f13-003999337e70.jpg)

emmberk replied:
I'm not quite sure where we differed, but I think it might be that I included that d=2*search radius in radians. Here's my work leading up to `final double lonDifference = Math.toDegrees(Math.asin(Math.sin(Math.toRadians(latDifference)) /  Math.cos(Math.toRadians(lat))));`:
![image](https://user-images.githubusercontent.com/8041246/36002799-b5ad4d48-0ce8-11e8-98a7-6040a18f407b.png)

emmberk replied:
![image](https://user-images.githubusercontent.com/8041246/36002888-0c8bba6e-0ce9-11e8-896a-7306739b784b.png)

emmberk replied:
![image](https://user-images.githubusercontent.com/8041246/36002910-234c1884-0ce9-11e8-8550-ddb01cd684d8.png)
![image](https://user-images.githubusercontent.com/8041246/36002922-295543ae-0ce9-11e8-9a0c-d7014fb8972f.png)
![image](https://user-images.githubusercontent.com/8041246/36002930-2d999f8c-0ce9-11e8-84a5-3cbf725e1cdb.png)
![image](https://user-images.githubusercontent.com/8041246/36002939-33a7ec12-0ce9-11e8-8546-44f25220cea6.png)

emmberk commented:
build now

emmberk commented:
build now

emmberk commented:
I added a new commit to use [Vicenty's formula](https://en.wikipedia.org/wiki/Vincenty%27s_formulae#Direct_Problem) from @clockard's suggestion. You can look at the changes to the unit test to see the difference in results for the two algorithms. I also fixed a few small sonarqube findings in OpenSearchParserImpl.

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchParserImpl.java:
```diff
@@ -338,50 +349,114 @@ private boolean hasParameter(String parameter, List<String> parameters) {
+        sigma = searchRadiusInMeters / (LENGTH_OF_SEMI_MINOR_AXIS_IN_METERS * A) + deltaSigma;
+
+        iterationCount++;
+      } while (Math.abs(sigma - sigmaPrime) > 1e-12);
```
`1e-12` is about 0.006mm. This probably does not need to be that exact, especially because the resulting `double`s are parsed to strings. The maximum `iterationCount` for the situations in the unit tests is only 4.

clockard replied:
Might want to make this error a little larger to reduce the number of iterations.

emmberk replied:
We could probably calculate what this should be to give us the most accurate decimal in the URL, but I don't think it's that important to spend the time to find the best allowable difference here. This bbox is only calculated once per search, and the iterations are very quick.

emmberk commented on catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/impl/OpenSearchParserImpl.java:
```diff
@@ -422,6 +500,10 @@ private boolean hasParameter(String parameter, List<String> parameters) {
   }
 
+  private static String createBboxParamString(double[] bboxCoords) {
+    return Arrays.stream(bboxCoords).mapToObj(Double::toString).collect(Collectors.joining(","));
```
I also added some asserts in the unit tests to make sure that the bbox parameter contains the values is the correct west,south,east,north order.

emmberk commented:
build now

emmberk commented:
build now

emmberk commented:
```
[ERROR] Failed to execute goal on project catalog-ui-search: Could not resolve dependencies for project ddf.ui:catalog-ui-search:bundle:2.11.6-SNAPSHOT: Could not transfer artifact org.geotools:gt-opengis:jar:14.4
...
```

emmberk commented:
build now

emmberk commented:
build now

emmberk commented:
```
[ERROR] Failed to execute goal on project klv: Could not resolve dependencies for project org.codice.ddf:klv:bundle:2.11.6-SNAPSHOT: Could not transfer artifact org.apache.poi:ooxml-schemas:jar:1.3
...
```

emmberk commented:
build now

emmberk commented:
```
[ERROR] Failed to execute goal on project klv: Could not resolve dependencies for project org.codice.ddf:klv:bundle:2.11.6-SNAPSHOT: Could not transfer artifact org.apache.poi:ooxml-schemas:jar:1.3
...
```

emmberk commented:
build now

clockard commented:
Manually verified this is working as expected.

emmberk commented:
```
[ERROR] Failed to execute goal on project rest-assured: Could not resolve dependencies for project ddf.thirdparty:rest-assured:bundle:2.11.6-SNAPSHOT: Could not transfer artifact org.codehaus.groovy:groovy:jar:2.4.4
...
```

emmberk commented:
build now
